README.md
# Hello World C++

**Note: This extension has only been tested on Ubuntu 20.04!**

## Features
Generate C++ Hello World project inside an open VSCode workspace. To use it, press `Ctrl+Shift+P`, then start typing `"Hello World C++"`. It will fill your folder with the necessary config files.

## Build and Run
- **Build:** Press `Ctrl+Shift+B`
- **Run:** Press `F5`

## Requirements
To build a C++ project, you have to install these build tools:
```
sudo apt install build-essential libtool autoconf unzip wget gdb libssl-dev
```

## Resources
VSCode C++ tutorial: https://code.visualstudio.com/docs/cpp/config-linux\
Download latest CMake: https://cmake.org/download/\
Install CMake: https://cmake.org/install/


## Generated Templates:
`.vscode/c_cpp_properties.json`
```
{
    "configurations": [
        {
            "name": "Linux",
            "includePath": [
                "${default}"
            ],
            "defines": [],
            "compilerPath": "/usr/bin/gcc",
            "cStandard": "gnu17",
            "cppStandard": "gnu++17",
            "intelliSenseMode": "linux-gcc-x64"
        }
    ],
    "version": 4
}
```

`.vscode/tasks.json`
```
{
    "version": "2.0.0",
    "tasks": [
        {
            "label": "Make build folder",
            "type": "shell",
            "linux": {
                "command": "mkdir -p ./build"
            }
        },
        {
            "label": "Create makefile",
            "type": "shell",
            "options": {
                "cwd": "${workspaceRoot}/build"
            },
            "command": "cmake -DCMAKE_CXX_COMPILER=/usr/bin/g++ ${workspaceRoot}",
            "dependsOn": [
                "Make build folder"
            ]
        },
        {
            "label": "Build",
            "type": "shell",
            "options": {
                "cwd": "${workspaceRoot}/build"
            },
            "command": "make",
            "dependsOn": [
                "Create makefile"
            ],
            "problemMatcher": [],
            "group": {
                "kind": "build",
                "isDefault": true
            }
        }
    ]
}
```

`.vscode/launch.json`
```
{
    "version": "0.2.0",
    "configurations": [
        {
            "name": "(gdb) Launch",
            "type": "cppdbg",
            "request": "launch",
            "program": "${workspaceFolder}/build/<project_name>",
            "stopAtEntry": false,
            "cwd": "${workspaceFolder}",
            "environment": [],
            "externalConsole": false,
            "MIMode": "gdb",
            "miDebuggerPath": "/usr/bin/gdb",
            "setupCommands": [
                {
                    "description": "Enable pretty-printing for gdb",
                    "text": "-enable-pretty-printing",
                    "ignoreFailures": true
                }
            ]
        }
    ]
}
```

`CMakeLists.txt`
```
cmake_minimum_required(VERSION 3.10)
project(<project_name>)

# add the executable
add_executable(\${PROJECT_NAME} main.cpp)
```
`main.cpp`
```
#include <iostream>

using namespace std;

int main() {
    cout << "Hello World!\\n";
}
```

CHANGELOG.md
# Change Log

All notable changes to the "cpp-builder" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Unreleased]

- Initial release

src/generator.ts
import * as vscode from 'vscode';
import * as fs from 'fs';
import * as path from 'path';

export class Generator {
    private wf: string | undefined = undefined;

    constructor() { }

    async execute(): Promise<void> {
        if (vscode.workspace.workspaceFolders !== undefined) {
            if (vscode.workspace.workspaceFolders.length > 1) {
                let workspaces: string[] = [];
                for (let i = 0; i < vscode.workspace.workspaceFolders.length; i++) {
                    workspaces.push(vscode.workspace.workspaceFolders[i].uri.path);
                }
                this.wf = await vscode.window.showQuickPick(workspaces, {});
            } else {
                this.wf = vscode.workspace.workspaceFolders[0].uri.path;
            }
        } else {
            vscode.window.showErrorMessage(
                `C++ Builder: Working folder not found, open a folder and try again`);
            return;
        }

        this.create();
    }

    async create() {
        if (this.wf !== undefined) {
            let properties = "c_cpp_properties.json";
            let tasks = "tasks.json";
            let launch = "launch.json";
            let main_cpp = this.wf + "/main.cpp";
            let cmake = this.wf + "/CMakeLists.txt";

            let basedir: string = this.wf + "/.vscode/";
            fs.mkdirSync(basedir, { recursive: true });

            const properties_content = `{
    "configurations": [
        {
            "name": "Linux",
            "includePath": [
                "\${default}"
            ],
            "defines": [],
            "compilerPath": "/usr/bin/gcc",
            "cStandard": "gnu17",
            "cppStandard": "gnu++17",
            "intelliSenseMode": "linux-gcc-x64"
        }
    ],
    "version": 4
}`

            const tasks_content = `{
    "version": "2.0.0",
    "tasks": [
        {
            "label": "Make build folder",
            "type": "shell",
            "linux": {
                "command": "mkdir -p ./build"
            }
        },
        {
            "label": "Create makefile",
            "type": "shell",
            "options": {
                "cwd": "\${workspaceRoot}/build"
            },
            "command": "cmake -DCMAKE_CXX_COMPILER=/usr/bin/g++ \${workspaceRoot}",
            "dependsOn": [
                "Make build folder"
            ]
        },
        {
            "label": "Build",
            "type": "shell",
            "options": {
                "cwd": "\${workspaceRoot}/build"
            },
            "command": "make",
            "dependsOn": [
                "Create makefile"
            ],
            "problemMatcher": [],
            "group": {
                "kind": "build",
                "isDefault": true
            }
        }
    ]
}`
            const launch_content = `{
    "version": "0.2.0",
    "configurations": [
        {
            "name": "(gdb) Launch",
            "type": "cppdbg",
            "request": "launch",
            "program": "\${workspaceFolder}/build/${path.basename(this.wf)}",
            "stopAtEntry": false,
            "cwd": "\${workspaceFolder}",
            "environment": [],
            "externalConsole": false,
            "MIMode": "gdb",
            "miDebuggerPath": "/usr/bin/gdb",
            "setupCommands": [
                {
                    "description": "Enable pretty-printing for gdb",
                    "text": "-enable-pretty-printing",
                    "ignoreFailures": true
                }
            ]
        }
    ]
}`

            const cmake_content = `cmake_minimum_required(VERSION 3.10)
project(${path.basename(this.wf)})

# add the executable
add_executable(\${PROJECT_NAME} main.cpp)`

            const main_cpp_content = `#include <iostream>

using namespace std;

int main() {
    cout << "Hello World!\\n";
}`

            try {
                fs.writeFileSync(basedir + properties, properties_content);
                fs.writeFileSync(basedir + tasks, tasks_content);
                fs.writeFileSync(basedir + launch, launch_content);

                if (!fs.existsSync(main_cpp)) {
                    fs.writeFileSync(main_cpp, main_cpp_content);
                } else {
                    const answer = await vscode.window.showWarningMessage(`${main_cpp} already exists. Do you want to override?`, "Yes", "No");
                    if (answer === "Yes") {
                        fs.writeFileSync(main_cpp, main_cpp_content);
                    }
                }
                
                if (!fs.existsSync(cmake)) {
                    fs.writeFileSync(cmake, cmake_content);
                }else {
                    const answer = await vscode.window.showWarningMessage(`${cmake} already exists. Do you want to override?`,"Yes", "No");
                    if (answer === "Yes") {
                        fs.writeFileSync(cmake, cmake_content);
                    }
                }
            } catch (err) {
                console.error(err);
            }
        }
    }
}